### PR TITLE
Use the Hash of the IP address instead of the actual IP address

### PR DESCRIPTION
**Pull request by cammellos**
For privacy reasons/compliance we want to keep the least amount
of information in our application.

https://discuss.status.im/t/status-app-privacy-policy-refresh/1762/3

This hashes the IP addresses before using them as key, so that the clear
IP address is never stored.

### CODE REVIEW

File: waku/common/rate_limiter.go [hunk 1]
```diff
@@ -186,6 +187,14 @@ func (r *PeerRateLimiter) Decorate(p RateLimiterPeer, rw p2p.MsgReadWriter, runL
...
 				// as IP() might return a nil value
 				ip = p.IP().String()
+
+				// If not the special <nil> case, take the hash of the IP address
+				// so it's not stored on our side
+				if ip != "<nil>" {
+					h := make([]byte, 64)
+					sha3.ShakeSum256(h, []byte(ip))
+					ip = fmt.Sprintf("%x", h)
+				}
```
Comment (jakubgs) [inline waku/common/rate_limiter.go:197]: Could we just make this into a `obfuscateIp` helper?
```go
func obfuscateIp(ip string) string {
	if ip == "<nil>" {
		return ip
	}
	h := make([]byte, 64)
	sha3.ShakeSum256(h, []byte(ip))
	return fmt.Sprintf("%x", h)
}
```

**Review by jakubgs — CHANGES_REQUESTED** [review]
Good idea, but anything that can be its own function should be.

File: waku/common/rate_limiter.go (hunk 1 above)
Comment (cammellos) [reply waku/common/rate_limiter.go:197]: Yes, can I just call it `hashIPAddress` instead?

Comment (jakubgs) [reply waku/common/rate_limiter.go:197]: whatever

**Review by Samyoul — APPROVED** [review]
Aside from addressing what @jakubgs has raised, this PR looks good.

File: waku/common/rate_limiter.go [hunk 2]
```diff
@@ -25,10 +25,11 @@ import (
 	"net"
 	"time"
 
-	"github.com/tsenart/tb"
+	"golang.org/x/crypto/sha3"
 
 	"github.com/ethereum/go-ethereum/p2p"
 	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/tsenart/tb"
```
Comment (Samyoul) [inline waku/common/rate_limiter.go:32]: Ah, I think that the linter is complaining that your imports aren't grouped by repo.

```suggestion
	"golang.org/x/crypto/sha3"

	"github.com/ethereum/go-ethereum/p2p"
	"github.com/ethereum/go-ethereum/p2p/enode"

	"github.com/tsenart/tb"
```

Comment (Samyoul) [reply waku/common/rate_limiter.go:32]: The linter gives the worst messages for this kind of linting issue.

**Comment by cammellos** [comment]
Is this even needed then? 
As I understand, this is just a box ticking exercise (it's trivial to build a list of the IP addresses hash), but technically we would not be storing IP addressing in memory. We can add a salt generated at runtime if needed, but in terms of privacy is identical (if an attacker gets access to the memory it will be able to access the hashes and the salt).

cc @hesterbruikman

**Comment by Samyoul** [comment]
> (if an attacker gets access to the memory it will be able to access the hashes and the salt

Would the attacker also have access to the private key(s) if memory access is compromised?

**Comment by cammellos** [comment]
> Would the attacker also have access to the private key(s) if memory access is compromised?

Yes, at least chat keys

**Comment by hesterbruikman** [comment]
Missed a lot of communication around this. Originally what I wanted to understand was:
>  I’d like to understand if rate limiting by IP address puts us in the controller category. Are IP addresses, stored, for what time?

If a hash of the IP address is stored (rather than the IP address itself). That might make a difference. I wasn't aware though that the IP address is only stored in memory, which in itself might make a difference. Let me verify both cases with the legal counsel. 

Objective is to see what we can do to not store any personal identifiable information (IP addresses). This a) avoids having to jump through legal hoops, reduce the number of boxes to tick, and b) pushes for what we can do to increase privacy. If hashes don't make a difference for `b`, `a` remains, pending on whether this makes a difference compared to storing in memory

File: waku/common/rate_limiter.go (hunk 2 above)
Comment (andremedeiros) [reply waku/common/rate_limiter.go:32]: I think what the linter wants is:

```
- Imports from stdlib

- Imports from 3rd party vendors

- Imports from your own project
```

**Comment by cyanlemons** [comment]
>We can add a salt generated at runtime if needed, but in terms of privacy is identical (if an attacker gets access to the memory it will be able to access the hashes and the salt).

@cammellos There is a slight benefit to adding a salt even if it gets compromised. You are immune to rainbow tables. SHA256 is rather common, and I wouldn't be surprised if someone already precomputed the hash of every single IPv4 address. If a salt is used, an attacker that gains access to the hashes would be required to recompute the full table every time.

**Pull request closed without merging**